In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

try:
    # Attempt to read the CSV with ISO-8859-1 encoding
    df = pd.read_csv('working-df/AviationData.csv', encoding='latin-1')
    print("Successfully read CSV with ISO-8859-1 encoding.")
except UnicodeDecodeError:
    try:
        # If ISO-8859-1 fails, attempt with cp1252 encoding
        df = pd.read_csv('working-df/AviationData.csv', encoding='cp1252')
        print("Successfully read CSV with cp1252 encoding.")
    except UnicodeDecodeError:
         try:
            #If cp1252 fails, attempt with utf-16 encoding
            df = pd.read_csv('working-df/AviationData.csv', encoding='utf-16')
            print("Successfully read CSV with utf-16 encoding.")
         except UnicodeDecodeError as e:
            print(f"Failed to read the CSV with ISO-8859-1, cp1252, and utf-16 encodings. Error: {e}")
            raise # Raise the exception so the user can see the error if none of the encodings work
df.info()

<ipython-input-2-e7125e57eb9f>:8: DtypeWarning: Columns (6,7,28) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('working-df/AviationData.csv', encoding='latin-1')


Successfully read CSV with ISO-8859-1 encoding.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  obj

In [3]:
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude  Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   
3       EUREKA, CA  United States        NaN        NaN          NaN   
4       Canton, OH  United States        NaN        NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

# Cleaning the Data
rename columns to remove dots as they may cause errors in Python (replace dots with dashes or underscores)

In [4]:
# replace dots in column names with underscores
df.columns = df.columns.str.replace('.', '_')

In [5]:
df.head()

Event_Id Investigation_Type Accident_Number  Event_Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude  Longitude Airport_Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   
3       EUREKA, CA  United States        NaN        NaN          NaN   
4       Canton, OH  United States        NaN        NaN          NaN   

  Airport_Name  ... Purpose_of_flight Air_carrier Total_Fatal_Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total_Serious_Injuries Total_Minor_Injuries Total_Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather_Condition  Broad_phase_of_flight   Report_Status Publication_Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

Check for duplicate rows


In [6]:
# check for duplicate rows in the Event_Id column
duplicate_rows_events = df[df.duplicated(subset=['Event_Id'], keep=False)]

duplicate_rows_events.head(10)

Event_Id Investigation_Type Accident_Number  Event_Date  \
117  20020917X01908           Accident     DCA82AA012B  1982-01-19   
118  20020917X01908           Accident     DCA82AA012A  1982-01-19   
153  20020917X02259           Accident     LAX82FA049A  1982-01-23   
158  20020917X02400           Accident     MIA82FA038B  1982-01-23   
159  20020917X02400           Accident     MIA82FA038A  1982-01-23   
160  20020917X02259           Accident     LAX82FA049B  1982-01-23   
242  20020917X02585           Accident     SEA82DA028A  1982-02-06   
244  20020917X02173           Accident     LAX82DA065B  1982-02-06   
245  20020917X02585           Accident     SEA82DA028B  1982-02-06   
248  20020917X02173           Accident     LAX82DA065A  1982-02-06   

               Location        Country Latitude Longitude Airport_Code  \
117        ROCKPORT, TX  United States      NaN       NaN          RKP   
118        ROCKPORT, TX  United States      NaN       NaN          RKP   
153     VICTORVILLE, CA  United States      NaN       NaN          NaN   
158  NEWPORT RICHEY, FL  United States      NaN       NaN          NaN   
159  NEWPORT RICHEY, FL  United States      NaN       NaN          NaN   
160     VICTORVILLE, CA  United States      NaN       NaN          NaN   
242         MEDFORD, OR  United States      NaN       NaN          MFR   
244        SAN JOSE, CA  United States      NaN       NaN          RHV   
245         MEDFORD, OR  United States      NaN       NaN          MFR   
248        SAN JOSE, CA  United States      NaN       NaN          RHV   

               Airport_Name  ...    Purpose_of_flight Air_carrier  \
117  ARANSAS COUNTY AIRPORT  ...             Personal         NaN   
118  ARANSAS COUNTY AIRPORT  ...  Executive/corporate         NaN   
153                     NaN  ...             Personal         NaN   
158                     NaN  ...             Personal         NaN   
159                     NaN  ...             Personal         NaN   
160                     NaN  ...             Personal         NaN   
242  MEDFORD-JACKSON COUNTY  ...   Aerial Application         NaN   
244           RIED HILLVIEW  ...             Personal         NaN   
245  MEDFORD-JACKSON COUNTY  ...             Personal         NaN   
248           RIED HILLVIEW  ...             Personal         NaN   

    Total_Fatal_Injuries Total_Serious_Injuries Total_Minor_Injuries  \
117                  3.0                    0.0                  0.0   
118                  3.0                    0.0                  0.0   
153                  2.0                    0.0                  4.0   
158                  0.0                    0.0                  0.0   
159                  0.0                    0.0                  0.0   
160                  2.0                    0.0                  4.0   
242                  0.0                    0.0                  0.0   
244                  0.0                    0.0                  0.0   
245                  0.0                    0.0                  0.0   
248                  0.0                    0.0                  0.0   

    Total_Uninjured Weather_Condition  Broad_phase_of_flight   Report_Status  \
117             0.0               IMC               Approach  Probable Cause   
118             0.0               IMC               Approach  Probable Cause   
153             0.0               VMC                Unknown  Probable Cause   
158             3.0               VMC                 Cruise  Probable Cause   
159             3.0               VMC               Approach  Probable Cause   
160             0.0               VMC                 Cruise  Probable Cause   
242             3.0               VMC                   Taxi  Probable Cause   
244             3.0               VMC               Standing  Probable Cause   
245             3.0               VMC                   Taxi  Probable Cause   
248             3.0               VMC                   Taxi  Probable C

I see here that though these duplicate rows do represent separate aircraft in multi-aircraft incidents, the injury and/or fatality numbers are combined. This would constitute duplicate numbers in certain columns that would render errors in the analysis.

So let's remove the duplicates from this subset.

In [7]:
# remove the duplicate rows using the Event_Id column
df = df.drop_duplicates(subset=['Event_Id'], keep='first')

In [8]:
# check for duplicates again in Event_Id column
duplicate_rows_events = df[df.duplicated(subset=['Event_Id'], keep=False)]

duplicate_rows_events.info()

<class 'pandas.core.frame.DataFrame'>
Index: 0 entries
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event_Id                0 non-null      object 
 1   Investigation_Type      0 non-null      object 
 2   Accident_Number         0 non-null      object 
 3   Event_Date              0 non-null      object 
 4   Location                0 non-null      object 
 5   Country                 0 non-null      object 
 6   Latitude                0 non-null      object 
 7   Longitude               0 non-null      object 
 8   Airport_Code            0 non-null      object 
 9   Airport_Name            0 non-null      object 
 10  Injury_Severity         0 non-null      object 
 11  Aircraft_damage         0 non-null      object 
 12  Aircraft_Category       0 non-null      object 
 13  Registration_Number     0 non-null      object 
 14  Make                    0 non-null      object 
 15  Model 

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 87951 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event_Id                87951 non-null  object 
 1   Investigation_Type      87951 non-null  object 
 2   Accident_Number         87951 non-null  object 
 3   Event_Date              87951 non-null  object 
 4   Location                87899 non-null  object 
 5   Country                 87729 non-null  object 
 6   Latitude                34212 non-null  object 
 7   Longitude               34203 non-null  object 
 8   Airport_Code            49484 non-null  object 
 9   Airport_Name            52031 non-null  object 
 10  Injury_Severity         86961 non-null  object 
 11  Aircraft_damage         84848 non-null  object 
 12  Aircraft_Category       32181 non-null  object 
 13  Registration_Number     86601 non-null  object 
 14  Make                    87888 non-null  obj

In [10]:
# check for duplicate rows in the Accident_Number column to verify there are no more duplicates
duplicate_rows_accidents = df[df.duplicated(subset=['Accident_Number'], keep=False)]

if duplicate_rows_accidents.empty:
  print("No duplicate rows found.")
else:
  print("Duplicate rows found.")

No duplicate rows found.


Remove certain columns that are mostly empty and would not contain data useful to the intended analysis.

I propose removing Latitude, Longitude, Schedule, and Air_carrier as those columns are mostly empty and would not contribute to the analysis.

In [11]:
# remove Latitude, Longitude, Schedule, and Air_carrier columns

df = df.drop(['Latitude', 'Longitude', 'Schedule', 'Air_carrier'], axis=1)

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 87951 entries, 0 to 88888
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event_Id                87951 non-null  object 
 1   Investigation_Type      87951 non-null  object 
 2   Accident_Number         87951 non-null  object 
 3   Event_Date              87951 non-null  object 
 4   Location                87899 non-null  object 
 5   Country                 87729 non-null  object 
 6   Airport_Code            49484 non-null  object 
 7   Airport_Name            52031 non-null  object 
 8   Injury_Severity         86961 non-null  object 
 9   Aircraft_damage         84848 non-null  object 
 10  Aircraft_Category       32181 non-null  object 
 11  Registration_Number     86601 non-null  object 
 12  Make                    87888 non-null  object 
 13  Model                   87859 non-null  object 
 14  Amateur_Built           87851 non-null  obj

## Aircraft_Category
The column for Aircraft_Category is also mostly empty, but that data could be useful. The business is after all seeking data related to airplanes specifically, so removing the column entirely would not work well and removing rows that are not airplanes and keeping those that are would be beneficial. Simply removing all rows that do not have a category entry would greatly reduce the number of total rows available for analysis, and most of those removed would likely be airplanes.

I would like to explore the idea of filling in as many of the missing values as I can. This could be done to some extent by making use of the Make column. For example, for all Aircraft_Category entries that are present, if the Make entry is Cessna, then the Aircraft_Category is invariably Airplane. So the missing category entries for Cessnas could reasonably be filled in. The same is true for other Makes like Sikorsky (Helicopter).

See below for an example:

In [12]:
# Show the Aircraft_Category values for Cessna in the Make column
df[df['Make'] == 'Cessna']['Aircraft_Category'].unique()

array([nan, 'Airplane'], dtype=object)

In [13]:
# Show how many nan Aircraft_Category values there are for Cessna
df[df['Make'] == 'Cessna']['Aircraft_Category'].isna().sum()

18344

In [14]:
# Show how many 'Airplane' Aircraft_Category values there are for Cessna
df[(df['Make'] == 'Cessna') & (df['Aircraft_Category'] == 'Airplane')]['Aircraft_Category'].count()

3581

So we can add another 18344 airplane entries to our data by filling in the missing value for Cessna in the Aircraft_Category column

In [15]:
# Show the Aircraft_Category values for Skiorsky in the Make column
df[df['Make'] == 'Sikorsky']['Aircraft_Category'].unique()

array(['Helicopter', nan], dtype=object)

In [16]:
# Show how many nan Aircraft_Category values there are for Sikorsky
df[df['Make'] == 'Sikorsky']['Aircraft_Category'].isna().sum()

128

In [17]:
# Fill in Aircraft_Category as 'Airplane' for Cessna
df.loc[df['Make'] == 'Cessna', 'Aircraft_Category'] = 'Airplane'

In [18]:
# Show the Aircraft_Category values for Cessna in the Make column
df[df['Make'] == 'Cessna']['Aircraft_Category'].unique()

array(['Airplane'], dtype=object)

In [19]:
# Show how many 'Airplane' Aircraft_Category values there are for Cessna
df[(df['Make'] == 'Cessna') & (df['Aircraft_Category'] == 'Airplane')]['Aircraft_Category'].count()

21925

So now, instead of only 3500 Cessna airplanes, we have almost 22000 entries, greatly increasing the verified airplane subset.

So here I will continue finding Makes that are airplanes only, and filling in the missing values.

In [20]:
# print a list of the unique values in the Make column
df['Make'].value_counts()

Make
Cessna           21925
Piper            11903
CESSNA            4914
Beech             4290
PIPER             2841
                 ...  
Geertz               1
Conrad Menzel        1
Blucher              1
Gideon               1
ROYSE RALPH L        1
Name: count, Length: 8202, dtype: int64

I realize here that I need to do some further cleaning of the Make column so Cessna and CESSNA (and other similar issues) are not separate values.

In [21]:
# Clean the Make column for misspellings, all caps issues, etc
replacement_dict = {'PIPER': 'Piper', 'CESSNA': 'Cessna', 'BOEING': 'Boeing', 'BEECH': 'Beech',
                    'BELL': 'Bell', 'ROBINSON': 'Robinson', 'MOONEY': 'Mooney', 'BELLANCA': 'Bellanca',
                    'ROBINSON HELICOPTER COMPANY': 'Robinson', 'AIR TRACTOR INC': 'Air Tractor',
                    'HUGHES': 'Hughes', 'SCHWEIZER': 'Schweizer', 'ROBINSON HELICOPTER': 'Robinson',
                    'Robinson Helicopters': 'Robinson', 'Robinson Helicopter Co.': 'Robinson' }

df['Make'] = df['Make'].replace(replacement_dict)

In [22]:
# Show complete list of unique values from Make column with valuecounts over 100
df['Make'].value_counts()[df['Make'].value_counts() > 100]

Make
Cessna                            26839
Piper                             14744
Beech                              5332
Bell                               2706
Boeing                             2652
Robinson                           1625
Mooney                             1322
Grumman                            1080
Bellanca                           1033
Hughes                              931
Air Tractor                         805
Schweizer                           772
Mcdonnell Douglas                   499
Aeronca                             479
Maule                               443
Champion                            426
De Havilland                        370
Aero Commander                      356
Stinson                             342
Aerospatiale                        334
Rockwell                            328
Luscombe                            316
Taylorcraft                         316
Hiller                              311
North American                      294
AIRBUS                              251
Douglas                             250
Enstrom                             247
Grumman American                    222
CIRRUS DESIGN CORP                  218
Ayres                               213
Ercoupe (eng & Research Corp.)      155
Sikorsky                            153
AERONCA                             149
MAULE                               144
Swearingen                          141
Lake                                136
Balloon Works                       135
Airbus Industrie                    135
Fairchild                           131
EMBRAER                             130
EUROCOPTER                          128
Pitts                               128
Mitsubishi                          126
Grumman-schweizer                   121
Waco                                114
Aviat                               112
Let                                 111
Lockheed                            111
Learjet                             109
Burkhart Grob                       109
Embraer                             106
Name: count, dtype: int64

In [84]:
makes_list = ['Piper', 'Beech', 'Bell', 'Boeing']
unique_categories_per_make = df[df['Make'].isin(makes_list)].groupby('Make')['Aircraft_Category'].unique()

print(unique_categories_per_make)

Make
Beech                   [nan, Airplane]
Bell        [Helicopter, nan, Airplane]
Boeing    [Airplane, nan, Powered-Lift]
Piper                   [nan, Airplane]
Name: Aircraft_Category, dtype: object
